packages/adapters-library/src/adapters/stargate/products/farm/stargateFarmAdapter.ts

```diff
@@ -0,0 +1,454 @@
+        : (lpStakingContract as LpStakingTime).eToken()
+    ).then((rewardTokenAddress) =>
+      getTokenMetadata(rewardTokenAddress, this.chainId, this.provider),
+    )
+
+    const poolLength = await lpStakingContract.poolLength()
+
+    const poolPromises: Promise<ProtocolToken<AdditionalMetadata>>[] = []
+
+    poolPromises.push(
+      ...Array.from({ length: Number(poolLength) }, async (_, i) => {
+        const { lpToken: protocolTokenAddress } =
+          await lpStakingContract.poolInfo(i)
+
+        const protocolToken = await getTokenMetadata(
+          protocolTokenAddress,
+          this.chainId,
+          this.provider,
+        )
+
+        return {
+          ...protocolToken,
+          poolIndex: i,
+          rewardToken: await rewardTokenPromise,
+          lpStakingType,
+          lpStakingAddress,
+        }
+      }),
+    )
+
+    if (lpStakingTimeMetisAddress) {
+      poolPromises.push(
+        (async () => {
+          const lpStakingTimeMetisContract = LpStakingTime__factory.connect(
+            lpStakingTimeMetisAddress,
+            this.provider,
+          )
+
+          const [{ lpToken: protocolTokenAddress }, rewardTokenAddress] =
+            await Promise.all([
+              lpStakingTimeMetisContract.poolInfo(0),
+              lpStakingTimeMetisContract.eToken(),
+            ])
+
+          const [protocolToken, rewardToken] = await Promise.all([
+            getTokenMetadata(protocolTokenAddress, this.chainId, this.provider),
+            getTokenMetadata(rewardTokenAddress, this.chainId, this.provider),
+          ])
+
+          return {
+            ...protocolToken,
+            poolIndex: 0,
+            rewardToken,
+            lpStakingType: 'LpStakingTime',
+            lpStakingAddress: lpStakingTimeMetisAddress,
+          }
+        })(),
+      )
+    }
+
+    return await Promise.all(poolPromises)
+  }
+
+  private async getProtocolTokenByAddress(
+    protocolTokenAddress: string,
+  ): Promise<ProtocolToken<AdditionalMetadata>> {
+    return this.helpers.getProtocolTokenByAddress({
+      protocolTokens: await this.getProtocolTokens(),
+      protocolTokenAddress,
+    })
+  }
+
+  async getPositions({
+    userAddress,
+    blockNumber,
+    protocolTokenAddresses,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const protocolTokens = await this.getProtocolTokens()
+
+    return await filterMapAsync(protocolTokens, async (protocolToken) => {
+      if (
+        protocolTokenAddresses &&
+        !protocolTokenAddresses.includes(protocolToken.address)
+      ) {
+        return undefined
+      }
+
+      const lpStakingContract =
+        protocolToken.lpStakingType === 'LpStaking'
+          ? LpStaking__factory.connect(
+              protocolToken.lpStakingAddress,
+              this.provider,
+            )
+          : LpStakingTime__factory.connect(
+              protocolToken.lpStakingAddress,
+              this.provider,
+            )
```

Comment:
This is fine for now but what stopped you doing a separate adapter?

Reply:
It was my initial thought, since it looked that some chains used one contract and some the other. So my plan was to make a common adapter and then extend for the very tiny different bits.

Then I saw that some chains (mainnet) used both depending on the pool.

Overall, it's a question of whether these ifs are annoying enough to justify a whole new adapter that will either have almost identical code or an abstraction with then several bits of code in protected functions reimplemented for each type.

No strong opinion either way, it's a grey area.